table/tables/partition.go

```diff
@@ -166,6 +173,43 @@ func generatePartitionExpr(tblInfo *model.TableInfo) (*PartitionExpr, error) {
 	}, nil
 }
 
+func generateHashPartitionExpr(tblInfo *model.TableInfo) (*PartitionExpr, error) {
+	var column *expression.Column
+	// The caller should assure partition info is not nil.
+	pi := tblInfo.GetPartitionInfo()
+	ctx := mock.NewContext()
+	partitionPruneExprs := make([]expression.Expression, 0, len(pi.Definitions))
+	var buf bytes.Buffer
+	dbName := model.NewCIStr(ctx.GetSessionVars().CurrentDB)
+	columns := expression.ColumnInfos2ColumnsWithDBName(ctx, dbName, tblInfo.Name, tblInfo.Columns)
+	schema := expression.NewSchema(columns...)
+	for i := 0; i < int(pi.Num); i++ {
+		fmt.Fprintf(&buf, "MOD(ABS(%s),(%d))=%d", pi.Expr, pi.Num, i)
+		exprs, err := expression.ParseSimpleExprsWithSchema(ctx, buf.String(), schema)
+		if err != nil {
+			// If it got an error here, ddl may hang forever, so this error log is important.
+			log.Error("wrong table partition expression:", errors.ErrorStack(err), buf.String())
+			return nil, errors.Trace(err)
+		}
+		partitionPruneExprs = append(partitionPruneExprs, exprs[0])
+		buf.Reset()
+	}
+	exprs, err := expression.ParseSimpleExprsWithSchema(ctx, pi.Expr, schema)
+	if err != nil {
+		// If it got an error here, ddl may hang forever, so this error log is important.
+		log.Error("wrong table partition expression:", errors.ErrorStack(err), pi.Expr)
+		return nil, errors.Trace(err)
+	}
+	if col, ok := exprs[0].(*expression.Column); ok {
+		column = col
+	}
+	return &PartitionExpr{
+		Column: column,
+		Expr:   exprs[0],
+		Ranges: partitionPruneExprs,
```

Comment:
The name `Ranges` is ambiguous, use pruneExprs instead?

Reply:
`Ranges` original is the ranges for range partition.
Here I save `partitionPruneExprs` to range partition just want to reuse range partition code in selection. So It's better to use the original name.

Reply:
I know, range partition still can use pruneExprs name. Why not use it？

Reply:
This PR is big enough, maybe we can refact this in another PR.

Reply:
Use range partition to implement hash partition is a trick. It will be rewrite sooner or later here.
So I think keep it `Ranges` is fine. @winkyao